src/System.Management.Automation/engine/debugger/debugger.cs

```diff
@@ -1257,10 +1257,19 @@ private CommandBreakpoint AddCommandBreakpoint(CommandBreakpoint breakpoint)
         private LineBreakpoint AddLineBreakpoint(LineBreakpoint breakpoint)
         {
             AddBreakpointCommon(breakpoint);
-            _pendingBreakpoints[breakpoint.Id] = breakpoint;
+            AddPendingBreakpoint(breakpoint);
+
             return breakpoint;
         }
 
+        private void AddPendingBreakpoint(LineBreakpoint breakpoint)
+        {
+            _pendingBreakpoints.AddOrUpdate(
+                breakpoint.Script,
+                new ConcurrentDictionary<int, LineBreakpoint> { [breakpoint.Id] = breakpoint },
```

Comment:
I wonder if using a delegate here would be cheaper, since currently every call to `AddPendingBreakpoint` requires us to allocate, construct and update a new `ConcurrentDictionary<int, LineBreakpoint>`

Reply:
I hope @nohwnd will share a test site he used so that we could measure and compare too.

Reply:
Are you talking about the "Add" part of that call, or about internals of how Func will allocate? Because the Add part should be called only once per file.

I was testing it like this: 
```
git clone https://github.com/pester/Pester.git
cd Pester
# ensure you are on v5.0 branch
git checkout v5.0
# you only need to do this once
.\build.ps1 -clean
# run Pester tests with code coverage
.\test.ps1 -ci -SkipPTests
```

Reply:
The steps worked ~40 min on my note. I didn't know it is that bad.

There is so much code generation in PerfView tracing, this is how PowerShell works internally, that no doubt we can never remove it. This greatly reduces the possibilities for optimizing this code.

Given this and the exceptional specifics of this scenario, I would no doubt prefer #13673.

Reply:
> Because the Add part should be called only once per file.

Right, except that there's nothing to call since it's not a delegate. That operation is performed for every call before passing it to the `AddOrUpdate()` method.

I'm saying that if you put the `new ConcurrentDictionary<>{ ... }` inside a delegate and then pass the delegate to `AddOrUpdate()`, the laziness might improve resource usage a bit.

Reply:
@rjmholt thanks, now I understand what you meant. 

@iSazonov Were you able to try it on a production build with this fix as well? Was there a difference? I am okay with using this or #13673, but it would be nice to have at least one of them merged already.

Reply:
>it would be nice to have at least one of them merged already.

I don’t understand what’s going on for a long time :-) It would seem that they should be most interested in improving PowerShell, but they are almost absent here.

> Were you able to try it on a production build with this fix as well? Was there a difference?

I tried to make measurements with PerfView, but it was unsuccessful - there is too much dynamic code there. It only made it clear that this approach would never be as effective as #13673. So much effort is already buried here that only if MSFT team finds your efforts helpful, then I will continue too.